File: src/Graphics/OpenGL.hs
module Graphics.OpenGL where


--GLFW3 + GLAD bindings using inline-c

import Foreign
import Foreign.C.Types
import Foreign.C.String
import Foreign.Ptr (Ptr,nullPtr)
import Control.Monad
import qualified Language.C.Inline as C
import qualified Language.C.Inline.Unsafe as UC
import Math.Linear.Mat
import qualified Math.Linear.Mat as Mat
import qualified Memory.MemBlock as Mem
import Math.Nat
import Data.Monoid ((<>))
import Foreign.Ptr(FunPtr, freeHaskellFunPtr)

type GLFWkeyfun = Ptr () -> CInt -> CInt -> CInt -> CInt -> IO ()
type GLFWframebuffersizefun = Ptr () -> CInt -> CInt -> IO()
type GLFWerrorfun = CInt -> CString -> IO ()

-- a "wrapper" import gives a factory for converting a Haskell function to a foreign function pointer
foreign import ccall "wrapper"
  wrapFrameBufferSizeCallbackFun :: GLFWframebuffersizefun -> IO (FunPtr GLFWframebuffersizefun)

foreign import ccall "wrapper"
  wrapKeyCallbackFun :: GLFWkeyfun -> IO (FunPtr GLFWkeyfun)

foreign import ccall "wrapper"
  wrapErrorCallbackFun :: GLFWerrorfun -> IO (FunPtr GLFWerrorfun)

C.context (C.baseCtx <> C.funCtx)

--C.include "<glfw3dll.h>"
C.include "<math.h>"
C.include "<glad.h>"
C.include "<glfw3.h>"
C.include "<stdio.h>"


c_GLFW_TRUE  = 1 :: Int
c_GLFW_FALSE = 0 :: Int
c_GL_COLOR_BUFFER_BIT = 0x00004000 :: Int
c_GL_VERTEX_SHADER = 35633 :: Int
c_GL_FRAGMENT_SHADER = 35632 :: Int
c_GL_COMPILE_STATUS = 35713 :: Int
c_GL_ARRAY_BUFFER = 34962 :: Int
c_GL_ELEMENT_ARRAY_BUFFER = 34963 :: Int
c_GL_STATIC_DRAW = 35044 :: Int
c_GL_FLOAT = 5126 :: Int
c_GL_UNSIGNED_INT = 5125 :: Int
c_GL_TRIANGLES = 4 :: Int
c_GL_LINES = 1 :: Int
c_GL_CURRENT_PROGRAM = 35725 :: Int



c_GLFW_KEY_SPACE              = 32 :: Int
c_GLFW_KEY_APOSTROPHE         = 39 :: Int
c_GLFW_KEY_COMMA              = 44 :: Int
c_GLFW_KEY_MINUS              = 45 :: Int
c_GLFW_KEY_PERIOD             = 46 :: Int
c_GLFW_KEY_SLASH              = 47 :: Int
c_GLFW_KEY_0                  = 48 :: Int
c_GLFW_KEY_1                  = 49 :: Int
c_GLFW_KEY_2                  = 50 :: Int
c_GLFW_KEY_3                  = 51 :: Int
c_GLFW_KEY_4                  = 52 :: Int
c_GLFW_KEY_5                  = 53 :: Int
c_GLFW_KEY_6                  = 54 :: Int
c_GLFW_KEY_7                  = 55 :: Int
c_GLFW_KEY_8                  = 56 :: Int
c_GLFW_KEY_9                  = 57 :: Int
c_GLFW_KEY_SEMICOLON          = 59 :: Int
c_GLFW_KEY_EQUAL              = 61 :: Int
c_GLFW_KEY_A                  = 65 :: Int
c_GLFW_KEY_B                  = 66 :: Int
c_GLFW_KEY_C                  = 67 :: Int
c_GLFW_KEY_D                  = 68 :: Int
c_GLFW_KEY_E                  = 69 :: Int
c_GLFW_KEY_F                  = 70 :: Int
c_GLFW_KEY_G                  = 71 :: Int
c_GLFW_KEY_H                  = 72 :: Int
c_GLFW_KEY_I                  = 73 :: Int
c_GLFW_KEY_J                  = 74 :: Int
c_GLFW_KEY_K                  = 75 :: Int
c_GLFW_KEY_L                  = 76 :: Int
c_GLFW_KEY_M                  = 77 :: Int
c_GLFW_KEY_N                  = 78 :: Int
c_GLFW_KEY_O                  = 79 :: Int
c_GLFW_KEY_P                  = 80 :: Int
c_GLFW_KEY_Q                  = 81 :: Int
c_GLFW_KEY_R                  = 82 :: Int
c_GLFW_KEY_S                  = 83 :: Int
c_GLFW_KEY_T                  = 84 :: Int
c_GLFW_KEY_U                  = 85 :: Int
c_GLFW_KEY_V                  = 86 :: Int
c_GLFW_KEY_W                  = 87 :: Int
c_GLFW_KEY_X                  = 88 :: Int
c_GLFW_KEY_Y                  = 89 :: Int
c_GLFW_KEY_Z                  = 90 :: Int
c_GLFW_KEY_LEFT_BRACKET       = 91 :: Int
c_GLFW_KEY_BACKSLASH          = 92 :: Int
c_GLFW_KEY_RIGHT_BRACKET      = 93 :: Int
c_GLFW_KEY_GRAVE_ACCENT       = 96 :: Int
c_GLFW_KEY_WORLD_1            = 161 :: Int
c_GLFW_KEY_WORLD_2            = 162 :: Int


c_GLFW_KEY_ESCAPE             = 256 :: Int

glfwInit :: IO Int
glfwInit = fromIntegral <$> [C.exp|int{glfwInit()}|]

glfwSetErrorCallback :: GLFWerrorfun -> IO (Ptr ())
glfwSetErrorCallback callback = do
  wrapped <- wrapErrorCallbackFun callback
  let ptr = castFunPtrToPtr wrapped :: Ptr ()
  [C.exp|void*{ glfwSetErrorCallback ( $(void* ptr) ) }|]


glfwCreateWindow :: Int -> Int -> String -> IO (Ptr ())
glfwCreateWindow width height name = let
         w :: CInt
         w = fromIntegral width

         h :: CInt
         h = fromIntegral height

         ion :: IO CString
         ion = newCString name
      in
         do
            n <- ion
            let !ret = [C.exp|void*{glfwCreateWindow($(int w), $(int h), $(char* n), NULL, NULL)}|]
            free n
            ret


glfwSetWindowShouldClose :: Ptr () -> Bool -> IO ()
glfwSetWindowShouldClose w should = do
  let c = fromIntegral (if should then c_GLFW_TRUE else c_GLFW_FALSE) :: CChar
  [C.exp|void{ glfwSetWindowShouldClose ( $(void* w), $(char c) )  }|]

--returns a FunPtr that should be freed after deinitialization of OpenGL
glfwSetFramebufferSizeCallback :: Ptr () -> GLFWframebuffersizefun -> IO (FunPtr GLFWframebuffersizefun)
glfwSetFramebufferSizeCallback window callback = do
  w <- wrapFrameBufferSizeCallbackFun callback --TODO needs manual freeing
  let w2 = castFunPtrToPtr w --TODO this cast is pretty bad
  [C.exp|void{ glfwSetFramebufferSizeCallback ( $(void* window),  $(void* w2) )  }|]
  pure w

--returns a FunPtr that should be freed after deinitialization of OpenGL
glfwSetKeyCallback :: Ptr () -> GLFWkeyfun -> IO (FunPtr GLFWkeyfun)
glfwSetKeyCallback window callback = do
  w <- wrapKeyCallbackFun callback --TODO needs manual freeing
  let w2 = castFunPtrToPtr w --TODO this cast is pretty bad
  [C.exp|void{ glfwSetKeyCallback ( $(void* window),  $(void* w2) )  }|]
  pure w

glfwTerminate :: IO ()
glfwTerminate = [C.exp|void{glfwTerminate()}|]

glfwMakeContextCurrent :: Ptr () -> IO ()
glfwMakeContextCurrent handle = do
    [C.exp|void{glfwMakeContextCurrent( $(void* handle) )}|]

glfwWindowShouldClose :: Ptr () -> IO Int
glfwWindowShouldClose handle = do
    fromIntegral <$> [C.exp|int{glfwWindowShouldClose( $(void* handle) )}|]

glGetUniformLocation :: Int -> String -> IO Int
glGetUniformLocation program name = do
  let c1 = fromIntegral program :: CInt
  c2 <- newCString name
  !res <- fromIntegral <$> [C.exp|int{glGetUniformLocation ($(int c1), $(char* c2))}|]
  free c2
  pure res

glUniform1i :: Int -> Int -> IO ()
glUniform1i uniform val = do
  let c1 = fromIntegral uniform :: CInt
  let c2 = fromIntegral val :: CInt
  [C.exp|void{glUniform1i ( $(int c1), $(int c2) ) }|]

glUniform1f :: Int -> Float -> IO ()
glUniform1f uniform val = do
  let c1 = fromIntegral uniform :: CInt
  let c2 = realToFrac val :: CFloat
  [C.exp|void{glUniform1f ( $(int c1), $(float c2) ) }|]

glUniform2f :: Int -> Float -> Float -> IO ()
glUniform2f uniform val1 val2 = do
  let c1 = fromIntegral uniform :: CInt
  let c2 = realToFrac val1 :: CFloat
  let c3 = realToFrac val2 :: CFloat
  [C.exp|void{glUniform2f ( $(int c1), $(float c2), $(float c3) ) }|]

glUniform3f :: Int -> Float -> Float -> Float -> IO ()
glUniform3f uniform val1 val2 val3 = do
  let c1 = fromIntegral uniform :: CInt
  let c2 = realToFrac val1 :: CFloat
  let c3 = realToFrac val2 :: CFloat
  let c4 = realToFrac val3 :: CFloat
  [C.exp|void{glUniform3f ( $(int c1), $(float c2), $(float c3), $(float c4) ) }|]

glUniform4f :: Int -> Float -> Float -> Float -> Float -> IO ()
glUniform4f uniform val1 val2 val3 val4 = do
  let c1 = fromIntegral uniform :: CInt
  let c2 = realToFrac val1 :: CFloat
  let c3 = realToFrac val2 :: CFloat
  let c4 = realToFrac val3 :: CFloat
  let c5 = realToFrac val4 :: CFloat
  [C.exp|void{glUniform4f ( $(int c1), $(float c2), $(float c3), $(float c4) , $(float c5)) }|]

glUniformMatrix4fv :: Int -> Bool -> Mat N4 N4 Float -> IO ()
glUniformMatrix4fv uniform transpose mat = do
  let c1 = fromIntegral uniform :: CInt
  let c2 = if transpose then 1 else 0 :: CChar
  mem <- Mem.new 16 :: IO (Mem.MemBlock CFloat)
  mem@(Mem.MemBlock _ ptr _) <- Mat.foreach (\ v mem -> Mem.add mem $ realToFrac v) mat mem
  !_ <- [C.exp|void{glUniformMatrix4fv ( $(int c1), 1, $(char c2),  $(float* ptr)) }|]
  Mem.delete mem

glGetIntegerv :: Int -> Ptr CInt -> IO ()
glGetIntegerv par out = do
  let c1 = fromIntegral par :: CInt
  [C.exp|void{glGetIntegerv ( $(int c1), $(int* out) )}|]

glClear :: Int -> IO ()
glClear bitField = do
    let f = (fromIntegral bitField) :: CUInt
    [C.exp|void{glClear( $(unsigned int f) )}|]

glClearColor :: Float -> Float -> Float -> Float -> IO ()
glClearColor r g b a =
    let f1 = realToFrac r
        f2 = realToFrac g
        f3 = realToFrac b
        f4 = realToFrac a
    in [C.exp|void{glClearColor( $(float f1), $(float f2), $(float f3), $(float f4) )}|]


glViewport :: Int -> Int -> Int -> Int -> IO ()
glViewport x y width height = do
  let c1 = fromIntegral x :: CInt
  let c2 = fromIntegral y :: CInt
  let c3 = fromIntegral width :: CInt
  let c4 = fromIntegral height :: CInt

  [C.exp|void{glViewport ($(int c1), $(int c2), $(int c3), $(int c4)) }|]

glCreateProgram :: IO Int
glCreateProgram =
    fromIntegral <$> [C.exp|int{glCreateProgram()}|]

glCreateShader :: Int -> IO Int
glCreateShader mode =
    let f = fromIntegral mode :: CInt
    in  fromIntegral <$> [C.exp|int{glCreateShader( $(int f) )}|]

glShaderSource :: Int -> Int -> String -> IO ()
glShaderSource handle count source = do
    let h = fromIntegral handle :: CInt
    let c = fromIntegral count :: CInt
    str <- newCString source
    _ <- [C.exp|void{glShaderSource( $(int h), $(int c), & $(char* str), NULL)}|]
    free str
    pure ()

glCompileShader :: Int -> IO ()
glCompileShader handle = do
    let h = fromIntegral handle :: CInt
    [C.exp|void{glCompileShader( $(int h) )}|]

glGetShaderiv :: Int -> Int -> Ptr CInt -> IO ()
glGetShaderiv shader mode result = do
    let s = fromIntegral shader :: CInt
    let m = fromIntegral mode :: CInt
    [C.exp|void{glGetShaderiv( $(int s), $(int m), $(int* result))}|]


glGenVertexArrays :: IO Int
glGenVertexArrays = do
    val <- malloc :: IO (Ptr CInt)
    [C.exp|void{glGenVertexArrays( 1, $(int* val) )}|]
    !result <- peek val :: IO CInt
    free val
    pure $ fromIntegral result

glGenBuffers :: IO Int
glGenBuffers = do
    val <- malloc :: IO (Ptr CInt)
    [C.exp|void{glGenBuffers( 1, $(int* val) )}|]
    !result <- peek val :: IO CInt
    free val
    pure $ fromIntegral result

glBindVertexArray :: Int -> IO ()
glBindVertexArray val = do
    let con = fromIntegral val :: CInt
    [C.exp|void{glBindVertexArray( $(int con) )}|]

glBindBuffer :: Int -> Int -> IO ()
glBindBuffer target buffer = do
    let con1 = fromIntegral target :: CInt
    let con2 = fromIntegral buffer :: CInt
    [C.exp|void{glBindBuffer( $(int con1) , $(int con2))}|]

glBufferData :: Int -> Int -> Ptr () -> Int -> IO ()
glBufferData buffer size dat usage = do
  let con1 = fromIntegral buffer :: CInt
  let con2 = fromIntegral size :: CInt
  let con3 = fromIntegral usage :: CInt
  [C.exp|void{glBufferData( $(int con1) , $(int con2), $(void* dat), $(int con3))}|]

glDeleteVertexArrays :: Int -> IO ()
glDeleteVertexArrays array = do
    let con1 = fromIntegral array :: CInt
    [C.exp|void{glDeleteVertexArrays(1, & $(int con1))}|]

glDeleteBuffers :: Int -> IO ()
glDeleteBuffers buffer = do
    let con1 = fromIntegral buffer :: CInt
    [C.exp|void{glDeleteBuffers(1, & $(int con1))}|]

glDrawElements :: Int -> Int -> Int -> Ptr () -> IO ()
glDrawElements mode count typee indices = do
    let con1 = fromIntegral mode :: CInt
    let con2 = fromIntegral count :: CInt
    let con3 = fromIntegral typee :: CInt
    [C.exp|void{glDrawElements( $(int con1), $(int con2), $(int con3), $(void* indices))}|]

glVertexAttribPointer :: Int -> Int -> Int -> Bool -> Int -> Integer -> IO ()
glVertexAttribPointer index size typee normalized stride pointer = do
    let con1 = fromIntegral index :: CInt
    let con2 = fromIntegral size :: CInt
    let con3 = fromIntegral typee :: CInt
    let con4 = if normalized then 1 else 0 :: CChar
    let con5 = fromIntegral stride :: CInt
    let con6 = fromIntegral pointer :: CLong
    [C.exp|void{glVertexAttribPointer( $(int con1), $(int con2), $(int con3), $(char con4), $(int con5), $(long con6))}|]

glEnableVertexAttribArray :: Int -> IO ()
glEnableVertexAttribArray array = do
    let con1 = fromIntegral array :: CInt
    [C.exp|void{glEnableVertexAttribArray($(int con1))}|]

glGetShaderInfoLog :: Int -> IO String
glGetShaderInfoLog shader = do
    let s = fromIntegral shader :: CInt
    let c = 512 :: CInt
    let allocated = (mallocBytes $ fromIntegral c) :: IO CString
    str <- allocated
    [C.exp|void{glGetShaderInfoLog( $(int s), $(int c), NULL, $(char* str))}|]
    let !res = peekCString str
    free str
    res



glAttachShader :: Int -> Int -> IO ()
glAttachShader program shader = do
    let p = fromIntegral program :: CInt
    let s = fromIntegral shader :: CInt
    [C.exp|void{glAttachShader( $(int p), $(int s))}|]

glLinkProgram :: Int -> IO ()
glLinkProgram program = do
    let p = fromIntegral program :: CInt
    [C.exp|void{glLinkProgram( $(int p) )}|]

glValidateProgram :: Int -> IO ()
glValidateProgram program = do
    let p = fromIntegral program :: CInt
    [C.exp|void{glValidateProgram( $(int p))}|]

glUseProgram :: Int -> IO ()
glUseProgram prog = do
    let p = fromIntegral prog :: CInt
    [C.exp|void{glUseProgram( $(int p) )}|]

glfwSwapBuffers :: Ptr () -> IO ()
glfwSwapBuffers handle = do
    [C.exp|void{glfwSwapBuffers(  $(void* handle) )}|]

glfwPollEvents :: IO ()
glfwPollEvents = [C.exp|void{glfwPollEvents()}|]

gladInit :: IO Int
gladInit = fromIntegral <$> [C.exp|int{gladLoadGLLoader((GLADloadproc)glfwGetProcAddress)}|]

File: src/Graphics/OpenGL.c

#include <math.h>

#include <glad.h>

#include <glfw3.h>

#include <stdio.h>

int inline_c_Graphics_OpenGL_0_fee3d34036aba97ee59677ca0af9323c9b725942() {
return (glfwInit());
}


void * inline_c_Graphics_OpenGL_1_da53656f9f5ac994ddd8f1b73d0f444fd524b1e6(void * ptr_inline_c_0) {
return ( glfwSetErrorCallback ( ptr_inline_c_0 ) );
}


void * inline_c_Graphics_OpenGL_2_db4f395c1c1ae6e1438f3ac6700dbf9f6bb439a3(int w_inline_c_0, int h_inline_c_1, char * n_inline_c_2) {
return (glfwCreateWindow(w_inline_c_0, h_inline_c_1, n_inline_c_2, NULL, NULL));
}


void inline_c_Graphics_OpenGL_3_48c4e70452272da5857dd7db35c0754b23f9f631(void * w_inline_c_0, char c_inline_c_1) {
 glfwSetWindowShouldClose ( w_inline_c_0, c_inline_c_1 )  ;
}


void inline_c_Graphics_OpenGL_4_d59867973c1e305220ce6f4efb5db32299e446f2(void * window_inline_c_0, void * w2_inline_c_1) {
 glfwSetFramebufferSizeCallback ( window_inline_c_0,  w2_inline_c_1 )  ;
}


void inline_c_Graphics_OpenGL_5_1fcf3a2a2eeb7ebf726eb2e10742c916cd30b457(void * window_inline_c_0, void * w2_inline_c_1) {
 glfwSetKeyCallback ( window_inline_c_0,  w2_inline_c_1 )  ;
}


void inline_c_Graphics_OpenGL_6_c3143be6bdd43eb718dd7dcc5dc93a9ca935cf8c(void * window_inline_c_0, void * w2_inline_c_1) {
 glfwSetMouseButtonCallback ( window_inline_c_0,  w2_inline_c_1 )  ;
}


void inline_c_Graphics_OpenGL_7_cb5bb105f88c9f974c67b3e218092c8e6b95d003() {
glfwTerminate();
}


void inline_c_Graphics_OpenGL_8_cd1d4d8d0ea13a15d371aff077011363df1b5602(void * handle_inline_c_0) {
glfwMakeContextCurrent( handle_inline_c_0 );
}


int inline_c_Graphics_OpenGL_9_fc7d1172a02d9d31e36213443ee1e2d590307170(void * handle_inline_c_0) {
return (glfwWindowShouldClose( handle_inline_c_0 ));
}


int inline_c_Graphics_OpenGL_10_5cdd4bceed5f9463a3a557b1ec7618bee5e1fc9e(int c1_inline_c_0, char * c2_inline_c_1) {
return (glGetUniformLocation (c1_inline_c_0, c2_inline_c_1));
}


void inline_c_Graphics_OpenGL_11_c3f2f962641bcbccdef514f82669784e577337d3(int c1_inline_c_0, int c2_inline_c_1) {
glUniform1i ( c1_inline_c_0, c2_inline_c_1 ) ;
}


void inline_c_Graphics_OpenGL_12_ef0d766241dee7a5ce95f74d7673deba6fa82fd5(int c1_inline_c_0, float c2_inline_c_1) {
glUniform1f ( c1_inline_c_0, c2_inline_c_1 ) ;
}


void inline_c_Graphics_OpenGL_13_bfc217a210d157312f3a295dbdc678473988274b(int c1_inline_c_0, float c2_inline_c_1, float c3_inline_c_2) {
glUniform2f ( c1_inline_c_0, c2_inline_c_1, c3_inline_c_2 ) ;
}


void inline_c_Graphics_OpenGL_14_90f1190e28306e9403041c5a715109e7b2e3b7d3(int c1_inline_c_0, float c2_inline_c_1, float c3_inline_c_2, float c4_inline_c_3) {
glUniform3f ( c1_inline_c_0, c2_inline_c_1, c3_inline_c_2, c4_inline_c_3 ) ;
}


void inline_c_Graphics_OpenGL_15_1019537cf2426f4f5f2e606f631ac31e1c36a54f(int c1_inline_c_0, float c2_inline_c_1, float c3_inline_c_2, float c4_inline_c_3, float c5_inline_c_4) {
glUniform4f ( c1_inline_c_0, c2_inline_c_1, c3_inline_c_2, c4_inline_c_3 , c5_inline_c_4) ;
}


void inline_c_Graphics_OpenGL_16_52c671ffdbc38fc2b1810c818a4912bf8fbe340f(int c1_inline_c_0, char c2_inline_c_1, float * ptr_inline_c_2) {
glUniformMatrix4fv ( c1_inline_c_0, 1, c2_inline_c_1,  ptr_inline_c_2) ;
}


void inline_c_Graphics_OpenGL_17_d581474ed01d5f0a353df057f9693f287e0fb4af(int c1_inline_c_0, int * out_inline_c_1) {
glGetIntegerv ( c1_inline_c_0, out_inline_c_1 );
}


void inline_c_Graphics_OpenGL_18_c09408afbf64a10ce5af32cff78ed333ec80b59c(unsigned f_inline_c_0) {
glClear( f_inline_c_0 );
}


void inline_c_Graphics_OpenGL_19_6785b2b671fbbc7963a43b6b141823c3b1eef9a0(float f1_inline_c_0, float f2_inline_c_1, float f3_inline_c_2, float f4_inline_c_3) {
glClearColor( f1_inline_c_0, f2_inline_c_1, f3_inline_c_2, f4_inline_c_3 );
}


void inline_c_Graphics_OpenGL_20_dd06983ca0907dc595c3f5f9bfec1997040a755f(int c1_inline_c_0, int c2_inline_c_1, int c3_inline_c_2, int c4_inline_c_3) {
glViewport (c1_inline_c_0, c2_inline_c_1, c3_inline_c_2, c4_inline_c_3) ;
}


int inline_c_Graphics_OpenGL_21_8fca97e75f7333aaecf73813182d58f886515fdb() {
return (glCreateProgram());
}


int inline_c_Graphics_OpenGL_22_31437bd0dd45dbf2a047723adae95c20e63aa197(int f_inline_c_0) {
return (glCreateShader( f_inline_c_0 ));
}


void inline_c_Graphics_OpenGL_23_1d1ff53ea5c36a571299453c0acd44419e7be270(int h_inline_c_0, int c_inline_c_1, char * str_inline_c_2) {
glShaderSource( h_inline_c_0, c_inline_c_1, & str_inline_c_2, NULL);
}


void inline_c_Graphics_OpenGL_24_1c2d04a548ea1fc949be1fd4b8bfefb016dd83a9(int h_inline_c_0) {
glCompileShader( h_inline_c_0 );
}


void inline_c_Graphics_OpenGL_25_2db0e9fdc59c3042060a6085adf716948e398568(int s_inline_c_0, int m_inline_c_1, int * result_inline_c_2) {
glGetShaderiv( s_inline_c_0, m_inline_c_1, result_inline_c_2);
}


void inline_c_Graphics_OpenGL_26_2e5aed9b74e58d2b032a9cf0c71cc976518803a4(int * val_inline_c_0) {
glGenVertexArrays( 1, val_inline_c_0 );
}


void inline_c_Graphics_OpenGL_27_75653890e37c5649f8dddd01d9d5d0f9e5e3d89a(int * val_inline_c_0) {
glGenBuffers( 1, val_inline_c_0 );
}


void inline_c_Graphics_OpenGL_28_8c30fd7bd4d66d6d41abace5ca47a76225503384(int con_inline_c_0) {
glBindVertexArray( con_inline_c_0 );
}


void inline_c_Graphics_OpenGL_29_e6e8d53087ac9c93d9866b6377699ecc42609b87(int con1_inline_c_0, int con2_inline_c_1) {
glBindBuffer( con1_inline_c_0 , con2_inline_c_1);
}


void inline_c_Graphics_OpenGL_30_7a44555dba3fa95a68cc85dc2b24123965cff1b1(int con1_inline_c_0, int con2_inline_c_1, void * dat_inline_c_2, int con3_inline_c_3) {
glBufferData( con1_inline_c_0 , con2_inline_c_1, dat_inline_c_2, con3_inline_c_3);
}


void inline_c_Graphics_OpenGL_31_6510368ae469e4c99c02dc8afd80b250ee075726(int con1_inline_c_0) {
glDeleteVertexArrays(1, & con1_inline_c_0);
}


void inline_c_Graphics_OpenGL_32_1aa7beba1c3f9b841f072f8744b16cb630ad4de6(int con1_inline_c_0) {
glDeleteBuffers(1, & con1_inline_c_0);
}


void inline_c_Graphics_OpenGL_33_b10a67c9a9e0bdccda15ad678d47581c17e45fa1(int con1_inline_c_0, int con2_inline_c_1, int con3_inline_c_2, void * indices_inline_c_3) {
glDrawElements( con1_inline_c_0, con2_inline_c_1, con3_inline_c_2, indices_inline_c_3);
}


void inline_c_Graphics_OpenGL_34_a970403317152714ae09c062e7a0fba997d0d862(int con1_inline_c_0, int con2_inline_c_1, int con3_inline_c_2, char con4_inline_c_3, int con5_inline_c_4, long con6_inline_c_5) {
glVertexAttribPointer( con1_inline_c_0, con2_inline_c_1, con3_inline_c_2, con4_inline_c_3, con5_inline_c_4, con6_inline_c_5);
}


void inline_c_Graphics_OpenGL_35_9c29fc60032e670f4f3808f8ddcb9417cfdd60b8(int con1_inline_c_0) {
glEnableVertexAttribArray(con1_inline_c_0);
}


void inline_c_Graphics_OpenGL_36_fb61f8eb2ad6739aa36017499eab9e9a61f017c6(int s_inline_c_0, int c_inline_c_1, char * str_inline_c_2) {
glGetShaderInfoLog( s_inline_c_0, c_inline_c_1, NULL, str_inline_c_2);
}


void inline_c_Graphics_OpenGL_37_4ca00bfce78e35a58ad5c4e7454713960665c087(int p_inline_c_0, int s_inline_c_1) {
glAttachShader( p_inline_c_0, s_inline_c_1);
}


void inline_c_Graphics_OpenGL_38_931441ebfcd0fc375e356e9a2be028d3bc7b347d(int p_inline_c_0) {
glLinkProgram( p_inline_c_0 );
}


void inline_c_Graphics_OpenGL_39_debca14f821ffe578fbd1a644d8c06a8f27ced98(int p_inline_c_0) {
glValidateProgram( p_inline_c_0);
}


void inline_c_Graphics_OpenGL_40_7286441cc5b156de0700ada62b1e2ed2515bc4b1(int p_inline_c_0) {
glUseProgram( p_inline_c_0 );
}


void inline_c_Graphics_OpenGL_41_3bddc9598e1a070bd32160807e3b06e91fe3d569(void * handle_inline_c_0) {
glfwSwapBuffers(  handle_inline_c_0 );
}


void inline_c_Graphics_OpenGL_42_f8d3648b35898cc511aa09659307d930effdb00c() {
glfwPollEvents();
}


int inline_c_Graphics_OpenGL_43_9fa711438a2f561807c3de22852f879532533523() {
return (gladLoadGLLoader((GLADloadproc)glfwGetProcAddress));
}

